### ISSUE TITLE
"'await' has no effect on the type of this expression" from inferred `void` and `Promise<void>` union

### ISSUE THREAD

**Issue by joshkel** [microsoft/TypeScript#55456]
### 🔎 Search Terms

"'await' has no effect on the type of this expression" , "promise\<void\>" 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about [Why are functions returning non-void assignable to function returning void?](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-returning-non-void-assignable-to-function-returning-void)

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.2.1-rc#code/MYewdgzgLgBGIgA4wLwwBQEpUD4YG8BfAbgChSBDCATzGBgDMBXOqAS3EYXQYBsKA5gC4YAIwS8AphTDZ8pGDFCRYkgB6TgTKJNSN+AmAH4YVWvSy4CMAPQAqGABMQMdYkmQ2AN10B3EABOANYwdjYwhDAi8EhkihS+FGyqGlo6WGSEpEA

### 💻 Code

```ts
const noop = () => {};

async function foo(flag: boolean) {
  const execute = flag ? async () => { /* do expensive work */ } : noop;
  await execute();
}
```

### 🙁 Actual behavior

The type of `execute` is inferred as `() => void`, so TypeScript displays a spurious warning on the `await execute();` line, "'await' has no effect on the type of this expression.(80007)".

### 🙂 Expected behavior

TypeScript infers the type as `() => void | () => Promise<void>`, and calling `await execute()` generates no warnings.

Based on the [FAQ](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-returning-non-void-assignable-to-function-returning-void), I would expect explicitly assigning `() => Promise<void>` to `() => void` to work, but I would not expect `() => void | () => Promise<void>` to be inferred as `() => void` and correct code then generating compiler diagnostics.

### Additional information about the issue

I found some possibly related issues: #44201, #43921, #49755

If this issue is actually a duplicate of one of those, I apologize.

**Comment by fatcerberus** [comment]
I believe this is unfortunate fallout due to subtype reduction.

**Comment by RyanCavanaugh** [comment]
Yeah, there's not terribly much we can do here. The best solution is to annotate `noop`:
```ts
const noop = (): undefined => {};
```
